components/src/screen/pcd8544/pcd8544.adb

```diff
@@ -0,0 +1,291 @@
+------------------------------------------------------------------------------
+--                                                                          --
+--                    Copyright (C) 2020, AdaCore                           --
+--                                                                          --
+--  Redistribution and use in source and binary forms, with or without      --
+--  modification, are permitted provided that the following conditions are  --
+--  met:                                                                    --
+--     1. Redistributions of source code must retain the above copyright    --
+--        notice, this list of conditions and the following disclaimer.     --
+--     2. Redistributions in binary form must reproduce the above copyright --
+--        notice, this list of conditions and the following disclaimer in   --
+--        the documentation and/or other materials provided with the        --
+--        distribution.                                                     --
+--     3. Neither the name of AdaCore nor the names of its                  --
+--        contributors may be used to endorse or promote products derived   --
+--        from this software without specific prior written permission.     --
+--                                                                          --
+--   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS    --
+--   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT      --
+--   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR  --
+--   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT   --
+--   HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, --
+--   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT       --
+--   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,  --
+--   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY  --
+--   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT    --
+--   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE  --
+--   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.   --
+--                                                                          --
+------------------------------------------------------------------------------
+
+package body PCD8544 is
+
+   procedure Chip_Select (This : PCD8544_Device; Enabled : in Boolean)
+   is
+   begin
+      if This.CS /= null then
+         if Enabled then
+            This.CS.Clear;
+         else
+            This.CS.Set;
+         end if;
+         This.Time.Delay_Microseconds (1);
```

Comment:
The datasheet specifies 100ns minimum transitions for most of the signals. HAL.Time doesn't do nanoseconds, so I used 1us. In practice, it works fine without most of the delays; the reset delay is important as I saw inconsistent behavior without it. I left the delays in because I can imagine running this driver on something very fast like one of the Xilinx Zynq chips where the GPIO can switch much faster.

For reference, the GPIO on STM32F4 can switch at a maximum of 84 MHz (11.9ns).

Reply:
I'd be happy to remove these delays if we can safely assume that users of the library have control over their GPIO clocks. I'm not sure this is a safe assumption though.

Reply:
It's better to remove the delays because we don't do that in any other driver.

If there is indeed a speed problem on a fast platforms we will find a solution for all the drivers at once.
For instance I can imagine a "timed GPIO" implementation that would introduced a delay in GPIO when needed.

```ada
   type Timed_GPIO (Pt       : not null HAL.GPIO.Any_GPIO_Point;
                    Time     : not null HAL.Time.Any_Delays;
                    Us_Delay : Natural)
   is new HAL.GPIO.GPIO_Point with null record;

   procedure Set (This : in out Timed_GPIO) is
   begin
      This.Time.Delay_Microseconds (This.Us_Delay);
      This.Pt.Set;
   end Set;
```